p2p/libp2p/discovery/kadDhtDiscoverer2.go

```diff
@@ -0,0 +1,245 @@
+
+	refreshInterval  time.Duration
+	randezVous       string
+	initialPeersList []string
+}
+
+// NewKadDhtPeerDiscoverer2 creates a new kad-dht discovery type implementation
+// initialPeersList can be nil or empty, no initial connection will be attempted, a warning message will appear
+func NewKadDhtPeerDiscoverer2(
+	refreshInterval time.Duration,
+	randezVous string,
+	initialPeersList []string) *KadDhtDiscoverer2 {
+
+	isListNilOrEmpty := len(initialPeersList) == 0
+
+	if isListNilOrEmpty {
+		log.Warn("nil or empty initial peers list provided to kad dht implementation. " +
+			"No initial connection will be2 done")
+	}
+
+	return &KadDhtDiscoverer2{
+		refreshInterval:  refreshInterval,
+		randezVous:       randezVous,
+		initialPeersList: initialPeersList,
+	}
+}
+
+// Bootstrap will start the bootstrapping new peers process
+func (kdd *KadDhtDiscoverer2) Bootstrap() error {
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
+
+	if kdd.kadDHT != nil {
+		return p2p.ErrPeerDiscoveryProcessAlreadyStarted
+	}
+
+	if kdd.contextProvider == nil {
+		return p2p.ErrNilContextProvider
+	}
+
+	return kdd.startDHT()
+}
+
+// UpdateRandezVous change the randezVous string, and restart the discovery with the new protocols
+func (kdd *KadDhtDiscoverer2) UpdateRandezVous(s string) error {
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
+
+	if s == kdd.randezVous {
+		return nil
+	}
+
+	err := kdd.stopDHT()
+	if err != nil {
+		log.Debug("Error wile stopping kad-dht discovery, skip", "error", err)
+	}
+	kdd.randezVous = s
+	return kdd.startDHT()
+}
+
+func (kdd *KadDhtDiscoverer2) protocols() []protocol.ID {
+	return []protocol.ID{
+		protocol.ID(fmt.Sprintf("%s/erd_%s", opts.ProtocolDHT, kdd.randezVous)),
+		protocol.ID(fmt.Sprintf("%s/erd", opts.ProtocolDHT)),
+		//TODO: to be removed once the seed is updated
+		opts.ProtocolDHT,
+	}
+}
+
+func (kdd *KadDhtDiscoverer2) startDHT() error {
+	ctx := kdd.contextProvider.Context()
+	h := kdd.contextProvider.Host()
+
+	ctxrun, cancel := context.WithCancel(ctx)
+	hd, err := NewHostDecorator(h, ctxrun, 3, time.Second)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	kademliaDHT, err := dht.New(ctx, hd, opts.Protocols(kdd.protocols()...))
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	go kdd.connectToInitialAndBootstrap(ctxrun)
+
+	kdd.kadDHT = kademliaDHT
+	kdd.refreshCancel = cancel
+	return nil
+}
+
+func (kdd *KadDhtDiscoverer2) stopDHT() error {
+	if kdd.refreshCancel == nil {
+		return nil
+	}
+
+	kdd.refreshCancel()
+	kdd.refreshCancel = nil
+
+	h := kdd.contextProvider.Host()
+
+	for _, p := range kdd.protocols() {
+		h.RemoveStreamHandler(p)
+	}
+
+	err := kdd.kadDHT.Close()
+
+	kdd.kadDHT = nil
+
+	return err
+}
+
+func (kdd *KadDhtDiscoverer2) connectToInitialAndBootstrap(ctx context.Context) {
+	chanStartBootstrap := kdd.connectToOnePeerFromInitialPeersList(
+		kdd.refreshInterval,
+		kdd.initialPeersList)
+
+	go func() {
+		<-chanStartBootstrap
+
+		go func() {
+			for {
```

Comment:
create a new method

Reply:
why is "for" needed ?

Reply:
Added TODO to refactor this. Will need some time to test the new implementation